File: apps/ios/MagicMobile/NativeDownloadsView.swift
import SwiftUI
import UIKit

struct NativeDownloadDeck: Identifiable {
    let id: String
    let name: String
    let cardNames: [String]

    init(id: String, deck: DeckList) {
        self.id = id
        name = deck.name
        cardNames = Array(Set(deck.entries.map(\.cardName) + [deck.commander?.cardName].compactMap { $0 })).sorted()
    }
}

/// Explicit artwork downloads, separate from the bundled rules engine and catalogue.
struct NativeDownloadsView: View {
    let decks: [NativeDownloadDeck]
    let engineReady: Bool
    @State private var selectedDeckID: String
    @AppStorage("magicmobile.artworkDownloadTokens") private var includeTokens = true
    @AppStorage("magicmobile.artworkDownloadScope") private var scope = "catalogue"
    @State private var catalogueNames: [String] = []
    @State private var loadingCatalogue = true
    @State private var catalogueError: String?
    @State private var confirmFullDownload = false
    @AppStorage("magicmobile.artworkDownloadQuality") private var quality: NativeArtworkQuality = .standard
    @ObservedObject private var downloads = NativeAssetDownloads.shared
    @AppStorage(NativeArtworkPreference.key) private var remoteArtwork = false
    @Environment(\.dismiss) private var dismiss

    init(decks: [NativeDownloadDeck], selectedDeckID: String, engineReady: Bool) {
        self.decks = decks
        self.engineReady = engineReady
        let previousDeck = MagicMobilePreferences.current.string(forKey: "magicmobile.artworkDownloadDeck") ?? selectedDeckID
        _selectedDeckID = State(initialValue: decks.contains { $0.id == previousDeck }
                               ? previousDeck : decks.first?.id ?? "")
    }

    private var names: [String] {
        switch scope {
        case "catalogue": return catalogueNames
        case "tokens": return []
        case "decks": return Array(Set(decks.flatMap(\.cardNames))).sorted()
        default: return decks.first { $0.id == selectedDeckID }?.cardNames ?? []
        }
    }
    private struct ScanSelection: Equatable {
        let scope: String
        let deck: String
        let quality: NativeArtworkQuality
        let catalogueCount: Int
        let running: Bool
    }
    @State private var scannedSelection: ScanSelection?
    private var scanSelection: ScanSelection {
        ScanSelection(scope: scope, deck: selectedDeckID, quality: quality, catalogueCount: catalogueNames.count, running: downloads.isRunning)
    }
    private var downloadsTokens: Bool { includeTokens || scope == "tokens" }
    private var missingCardCount: Int { scope == "tokens" ? 0 : downloads.missingNames.count }
    private var missingTokenCount: Int { downloadsTokens ? downloads.missingTokenNames.count : 0 }
    private var scanPending: Bool { downloads.isScanning || !downloads.scanSucceeded || scannedSelection != scanSelection }
    private var discoveryIncomplete: Bool {
        scanPending || (scope == "catalogue" && (loadingCatalogue || downloads.faceDiscoveryPending)) ||
            (downloadsTokens && downloads.tokenDiscoveryRemaining > 0)
    }
    private var missingSummary: String {
        if scanPending { return "Checking needed to count missing artwork." }
        let cards = "\(missingCardCount.formatted()) card/face \(missingCardCount == 1 ? "image" : "images")"
        let tokens = "\(missingTokenCount.formatted()) token \(missingTokenCount == 1 ? "image" : "images")"
        let count = scope == "tokens" ? tokens : (downloadsTokens ? "\(cards) · \(tokens)" : cards)
        return discoveryIncomplete ? "Found so far: \(count). Checking needed for the remaining artwork." : "Missing: \(count)"
    }
    private var estimatedAdditionalSize: String {
        guard !discoveryIncomplete else { return "Checking needed" }
        let tokensToDownload = downloadsTokens ? downloads.downloadableMissingTokenCount : 0
        let bytes = Int64(missingCardCount + tokensToDownload) * Int64(quality.estimatedBytes)
        return "≈ \(ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file))"
    }
    private var preparingDownload: Bool {
        downloads.isScanning || downloads.status.hasPrefix("Preparing") || downloads.status.hasPrefix("Checking")
    }
    private func startDownload() {
        downloads.download(names: names, includeTokens: downloadsTokens, allowNetwork: remoteArtwork,
                           quality: quality, fullCatalogue: scope == "catalogue" || scope == "tokens", tokenOnly: scope == "tokens")
    }
    private var catalogueIncluded: Bool {
        Bundle.main.url(forResource: "ondevice-catalogue", withExtension: "json") != nil
    }

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    Picker("Download", selection: $scope) {
                        Text("Full catalogue · recommended").tag("catalogue")
                        Text("All supported tokens").tag("tokens")
                        Text("All saved & included decks").tag("decks")
                        Text("One deck").tag("deck")
                    }
                    .disabled(downloads.isRunning)
                    .accessibilityIdentifier("downloads.scope")
                    if scope == "deck" {
                        Picker("Deck", selection: $selectedDeckID) {
                            ForEach(decks) { Text($0.name).tag($0.id) }
                        }
                        .disabled(downloads.isRunning)
                        .accessibilityIdentifier("downloads.deck")
                    }
                    Picker("Image quality", selection: $quality) {
                        ForEach(NativeArtworkQuality.allCases) { Text($0.label).tag($0) }
                    }
                    .disabled(downloads.isRunning)
                    .accessibilityIdentifier("downloads.quality")
                    if scope != "tokens" { Toggle("Include tokens", isOn: $includeTokens)
                        .disabled(downloads.isRunning)
                    }
                } header: { Text("Artwork") } footer: {
                    Text(scope == "tokens" ? "Token-only downloads contain no ordinary card images." : "Full catalogue includes opponents’ cards too.")
                }

                Section("On this device") {
                    if loadingCatalogue && scope == "catalogue" { ProgressView("Reading the installed catalogue…") }
                    if let catalogueError, scope == "catalogue" { Text(catalogueError).foregroundStyle(.red) }
                    if scope != "tokens" {
                        LabeledContent("Cards", value: scanPending || (scope == "catalogue" && loadingCatalogue)
                                       ? "Checking needed" : "\(downloads.cardStored.formatted()) / \(downloads.cardTotal.formatted())")
                            .accessibilityIdentifier("downloads.cards")
                    }
                    LabeledContent("Tokens", value: scanPending || downloads.tokenDiscoveryRemaining > 0
                                   ? "Checking needed" : "\(downloads.tokenStored.formatted()) / \(downloads.tokenTotal.formatted())")
                    if downloads.isScanning { ProgressView("Checking local files…") }
                    LabeledContent("Stored", value: ByteCountFormatter.string(fromByteCount: Int64(downloads.storedBytes), countStyle: .file))
                    LabeledContent("Missing artwork", value: missingSummary)
                    LabeledContent("Estimated additional download", value: estimatedAdditionalSize)
                    Text("Approximate at \(quality.label.lowercased()) quality. Actual download and device storage vary with image sizes, metadata and images already stored.")
                        .font(.caption).foregroundStyle(.secondary)
                    Button("Check for missing artwork") {
                        Task { await downloads.scan(names: names, quality: quality, fullCatalogue: scope == "catalogue" || scope == "tokens", tokenOnly: scope == "tokens") }
                    }
                    .disabled(downloads.isRunning || downloads.isScanning)
                    .accessibilityIdentifier("downloads.check")
                }

                Section {
                    Toggle("Download card artwork", isOn: $remoteArtwork)
                        .accessibilityIdentifier("nativeArtwork.downloads")
                    Text("Uses Scryfall. Online requests share your IP and card names, including your hand.")
                        .font(.caption).foregroundStyle(.secondary)
                    if downloads.isRunning {
                        if preparingDownload {
                            ProgressView("Preparing image list…")
                        } else {
                            ProgressView(value: Double(downloads.completed), total: Double(max(1, downloads.total)))
                        }
                        Text(downloads.status).font(.callout)
                            .accessibilityIdentifier("downloads.status")
                        Button("Cancel download", role: .cancel) { downloads.cancel() }
                            .accessibilityIdentifier("downloads.cancel")
                    } else {
                        if downloads.total > 0 && !downloads.status.isEmpty {
                            Text(downloads.status).font(.callout)
                                .accessibilityIdentifier("downloads.status")
                        }
                        Button("Download missing artwork") {
                            if scope == "catalogue" || scope == "tokens" { confirmFullDownload = true }
                            else { startDownload() }
                        }
                        .disabled(!remoteArtwork || (names.isEmpty && scope != "tokens") || downloads.isScanning || (scope == "catalogue" && loadingCatalogue))
                        .accessibilityIdentifier("downloads.start")
                    }
                } header: { Text("Download") } footer: {
                    Text("You can play or leave the app while images download. Wi-Fi is recommended.")
                }

                Section {
                    DisclosureGroup("More info") {
                        Label(engineReady ? "Local engine ready" : "Local engine not ready", systemImage: engineReady ? "checkmark.circle" : "exclamationmark.circle")
                        Label(catalogueIncluded ? "Card catalogue included" : "Card catalogue unavailable", systemImage: catalogueIncluded ? "checkmark.circle" : "exclamationmark.circle")
                        Label("Mana symbols included", systemImage: "checkmark.circle")
                        Text("These downloads supply artwork for decks and games. Rules and the supported card catalogue are already included; artwork is optional.")
                        Text("Compact saves space. Standard balances clarity and size. High gives the sharpest inspection images. Higher-quality files already stored count toward lower-quality coverage.")
                        Text("Full catalogue covers this build’s supported cards, not every printing. Alternate faces are checked during download. Estimates use currently discovered missing images; more faces or tokens may be found while preparing. Actual download and storage vary. Check for missing artwork after app updates.")
                        Text("Full and token-only downloads use Scryfall’s bulk image index. Deck and on-demand requests share card names and your IP address. Stored artwork works offline.")
                        Text("Compact is fastest. Downloads use several direct image transfers at once and remember completed files. iOS controls background timing; force-quitting pauses transfers until you reopen the app. The initial image-list preparation may need the app open on a slow connection.")
                        Text("Storage is capped at 20 GB, with 1 GB of free space reserved. Unavailable or ambiguous token art remains a labeled placeholder. Use Download missing artwork to retry interruptions.")
                    }
                    .font(.callout)
                    .accessibilityIdentifier("downloads.info")
                }

                if !downloads.failures.isEmpty {
                    Section("Needs attention") {
                        ForEach(Array(downloads.failures.prefix(20).enumerated()), id: \.offset) { _, failure in
                            Text(failure).font(.callout)
                        }
                        if downloads.failures.count > 20 {
                            Text("And \(downloads.failures.count - 20) more issues. Check missing artwork below.")
                                .font(.caption).foregroundStyle(.secondary)
                        }
                        Text("Use Download missing artwork to retry. Already stored artwork is preserved.")
                            .font(.caption).foregroundStyle(.secondary)
                    }
                }
                if !downloads.missingTokenNames.isEmpty {
                    Section {
                        DisclosureGroup("Missing tokens · \(downloads.missingTokenNames.count.formatted())") {
                            ForEach(Array(downloads.missingTokenNames.prefix(20).enumerated()), id: \.offset) { _, name in Text(name) }
                        }
                    }
                }
                if !downloads.missingNames.isEmpty {
                    Section {
                        DisclosureGroup("Missing cards · \(downloads.missingNames.count.formatted())") {
                            ForEach(downloads.missingNames.prefix(20), id: \.self) { Text($0) }
                            if downloads.missingNames.count > 20 {
                                Text("And \(downloads.missingNames.count - 20) more")
                                    .foregroundStyle(.secondary)
                            }
                        }
                    }
                }
            }
            .scrollContentBackground(.hidden)
            .background(CommanderPresentation.canvas)
            .tint(CommanderPresentation.accent)
            .navigationTitle("Downloads")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar { ToolbarItem(placement: .confirmationAction) { Button("Done") { dismiss() } } }
            .task {
                do {
                    catalogueNames = try await Task.detached(priority: .utility) {
                        try NativeDeckMetadataCatalogue.bundled().artworkCardNames
                    }.value
                } catch { catalogueError = "The installed card catalogue could not be read. Deck downloads are still available." }
                loadingCatalogue = false
            }
            .task(id: scanSelection) {
                let selection = scanSelection
                if !downloads.isRunning {
                    await downloads.scan(names: names, quality: quality, fullCatalogue: scope == "catalogue" || scope == "tokens", tokenOnly: scope == "tokens")
                    if selection == scanSelection && !downloads.isScanning && downloads.scanSucceeded { scannedSelection = selection }
                }
            }
            .alert("Download missing artwork?", isPresented: $confirmFullDownload) {
                Button("Download missing images · \(quality.label)") { startDownload() }
                Button("Cancel", role: .cancel) {}
            } message: {
                Text("\(missingSummary) \(discoveryIncomplete ? "The additional download estimate needs checking." : "Estimated additional download: \(estimatedAdditionalSize).") This is approximate; actual download and device storage vary. Images continue downloading while you play or leave the app. Wi-Fi is recommended.")
            }
            .onChange(of: remoteArtwork) { _, enabled in if !enabled { downloads.cancel() } }
            .onChange(of: selectedDeckID) { _, deck in
                MagicMobilePreferences.current.set(deck, forKey: "magicmobile.artworkDownloadDeck")
            }
        }
        .preferredColorScheme(.dark)
    }
}

File: apps/android/app/src/main/java/io/magicmobile/android/ondevice/NativeDownloadsView.kt
package io.magicmobile.android.ondevice

import android.content.Context
import android.text.format.Formatter
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import io.magicmobile.android.Artwork
import io.magicmobile.android.ArtworkDownloadClient
import io.magicmobile.android.ArtworkDownloadService
import io.magicmobile.android.ArtworkQuality
import io.magicmobile.android.DownloadScan
import io.magicmobile.android.board.MenuEntry
import io.magicmobile.android.core.Deck
import io.magicmobile.android.ui.AppPreferences
import io.magicmobile.android.ui.BrandTheme
import io.magicmobile.android.ui.IosAlert
import io.magicmobile.android.ui.IosMenuPicker
import io.magicmobile.android.ui.IosSheetHeader
import io.magicmobile.android.ui.IosToggle
import io.magicmobile.android.ui.SfImage
import io.magicmobile.android.ui.SfWeight
import io.magicmobile.android.ui.rgb
import io.magicmobile.android.ui.sf
import kotlinx.coroutines.launch
import java.text.NumberFormat

/** One deck the downloads screen can target (NativeDownloadDeck). */
data class NativeDownloadDeck(val id: String, val name: String, val cardNames: List<String>) {
    companion object {
        fun of(id: String, deck: Deck) = NativeDownloadDeck(id, deck.name, deck.entries.map { it.name }.toSortedSet().toList())
    }
}

private val formRow = rgb(0.11, 0.11, 0.12)
private val secondary = Color.White.copy(alpha = 0.6f)

@Composable
private fun FormSection(header: String? = null, footer: String? = null, content: @Composable ColumnScope.() -> Unit) {
    Column(Modifier.fillMaxWidth()) {
        header?.let { Text(it.uppercase(), Modifier.padding(start = 16.dp, bottom = 6.dp), color = secondary, style = sf(13f)) }
        Column(Modifier.fillMaxWidth().background(formRow, RoundedCornerShape(10.dp)).padding(horizontal = 16.dp, vertical = 4.dp), content = content)
        footer?.let { Text(it, Modifier.padding(start = 16.dp, end = 16.dp, top = 6.dp), color = secondary, style = sf(13f)) }
    }
}

@Composable
private fun RowDivider() = HorizontalDivider(thickness = 0.5.dp, color = Color.White.copy(alpha = 0.12f))

@Composable
private fun Labeled(title: String, value: String, modifier: Modifier = Modifier) {
    Row(modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp).padding(vertical = 10.dp), verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(title, color = Color.White, style = sf(17f))
        Text(value, Modifier.weight(1f), color = secondary, style = sf(17f), textAlign = androidx.compose.ui.text.style.TextAlign.End)
    }
}

@Composable
private fun PickerRow(title: String, value: String, enabled: Boolean, entries: () -> List<MenuEntry>, label: String) {
    Row(Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp).alpha(if (enabled) 1f else 0.45f), verticalAlignment = Alignment.CenterVertically) {
        Text(title, color = Color.White, style = sf(17f), maxLines = 1, softWrap = false)
        Box(Modifier.weight(1f).padding(start = 12.dp), contentAlignment = Alignment.CenterEnd) { IosMenuPicker(value, entries, enabled = enabled, label = label) }
    }
}

@Composable
private fun FormButton(title: String, enabled: Boolean, label: String, destructive: Boolean = false, onClick: () -> Unit) {
    Box(Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp).clickable(enabled = enabled, onClick = onClick).semantics { contentDescription = label },
        contentAlignment = Alignment.CenterStart) {
        Text(title, color = (if (destructive) rgb(1.0, 0.27, 0.23) else BrandTheme.ember).copy(alpha = if (enabled) 1f else 0.35f), style = sf(17f))
    }
}

/** NativeDownloadsView.swift: explicit artwork downloads, separate from the bundled rules engine and catalogue. */
@Composable
fun NativeDownloadsView(decks: List<NativeDownloadDeck>, selectedDeckID: String, engineReady: Boolean, catalogueNames: List<String>?, catalogueError: String?,
                        dismiss: () -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val number = remember { NumberFormat.getIntegerInstance() }
    var includeTokens by AppPreferences.boolean("magicmobile.artworkDownloadTokens", true)
    var downloadScope by AppPreferences.string("magicmobile.artworkDownloadScope", "catalogue")
    var qualityID by AppPreferences.string("magicmobile.artworkDownloadQuality", "standard")
    var savedDeck by AppPreferences.string("magicmobile.artworkDownloadDeck", selectedDeckID)
    val quality = ArtworkQuality.entries.firstOrNull { it.id == qualityID } ?: ArtworkQuality.STANDARD
    val deckID = if (decks.any { it.id == savedDeck }) savedDeck else decks.firstOrNull()?.id ?: ""
    val consentRevision = Artwork.consentRevision
    val remoteArtwork = remember(consentRevision) { Artwork.enabled(context) }
    val downloadState by ArtworkDownloadService.state.collectAsState()
    val running = downloadState.running
    var scanning by remember { mutableStateOf(false) }
    var scan by remember { mutableStateOf<DownloadScan?>(null) }
    var scanned by remember { mutableStateOf<String?>(null) }
    var confirmFull by remember { mutableStateOf(false) }
    val notificationPermission = rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) {}
    val loadingCatalogue = catalogueNames == null && catalogueError == null
    val names = when (downloadScope) {
        "catalogue" -> catalogueNames ?: emptyList()
        "tokens" -> emptyList()
        "decks" -> decks.flatMap { it.cardNames }.toSortedSet().toList()
        else -> decks.firstOrNull { it.id == deckID }?.cardNames ?: emptyList()
    }
    val downloadsTokens = includeTokens || downloadScope == "tokens"
    val selectionKey = "$downloadScope|$deckID|${quality.id}|${names.size}|$downloadsTokens"
    val scanPending = scanning || scan == null || scanned != selectionKey
    val current = scan
    val missingCards = if (downloadScope == "tokens" || current == null) 0 else current.missingCards.size
    val missingTokens = if (!downloadsTokens || current == null) 0 else current.missingTokens.size
    val discoveryIncomplete = scanPending || (downloadScope == "catalogue" && loadingCatalogue) || current?.coverageKnown != true
    val missingSummary = if (scanPending) "Checking needed to count missing artwork." else {
        val cards = "${number.format(missingCards)} card/face ${if (missingCards == 1) "image" else "images"}"
        val tokens = "${number.format(missingTokens)} token ${if (missingTokens == 1) "image" else "images"}"
        val count = if (downloadScope == "tokens") tokens else if (downloadsTokens) "$cards · $tokens" else cards
        if (discoveryIncomplete) "Found so far: $count. Checking needed for the remaining artwork." else "Missing: $count"
    }
    val estimate = if (discoveryIncomplete) "Checking needed" else "≈ ${Formatter.formatFileSize(context, (missingCards + missingTokens).toLong() * quality.estimatedBytes)}"
    val fullCatalogue = downloadScope == "catalogue" || downloadScope == "tokens"

    fun runScan() {
        val key = selectionKey
        if (downloadScope == "catalogue" && loadingCatalogue) return
        scanning = true
        scope.launch {
            runCatching { ArtworkDownloadClient(context).scan(names, quality, downloadsTokens) }
                .onSuccess { scan = it; scanned = key }
            scanning = false
        }
    }
    fun startDownload() {
        if (running || (names.isEmpty() && downloadScope != "tokens")) return
        val preferences = context.getSharedPreferences("magicmobile.artwork", Context.MODE_PRIVATE)
        if (android.os.Build.VERSION.SDK_INT >= 33 && !preferences.getBoolean("askedDownloadNotifications", false)) {
            preferences.edit().putBoolean("askedDownloadNotifications", true).apply()
            notificationPermission.launch(android.Manifest.permission.POST_NOTIFICATIONS)
        }
        scope.launch { runCatching { ArtworkDownloadService.start(context.applicationContext, names, quality, downloadsTokens, fullCatalogue) } }
    }
    LaunchedEffect(selectionKey, loadingCatalogue) { runScan() }
    LaunchedEffect(running) { if (!running) runScan() }

    Column(Modifier.fillMaxWidth().height((LocalConfiguration.current.screenHeightDp - 56).dp)) {
        IosSheetHeader("Downloads", dismiss)
        Column(Modifier.verticalScroll(rememberScrollState()).padding(horizontal = 16.dp).padding(bottom = 32.dp), verticalArrangement = Arrangement.spacedBy(24.dp)) {
            FormSection("Artwork", if (downloadScope == "tokens") "Token-only downloads contain no ordinary card images." else "Full catalogue includes opponents’ cards too.") {
                val scopes = listOf("catalogue" to "Full catalogue · recommended", "tokens" to "All supported tokens", "decks" to "All saved & included decks", "deck" to "One deck")
                PickerRow("Download", scopes.first { it.first == downloadScope }.second, !running, {
                    scopes.map { (value, title) -> MenuEntry.Item(title, checked = value == downloadScope) { downloadScope = value } }
                }, "downloads.scope")
                if (downloadScope == "deck") {
                    RowDivider()
                    PickerRow("Deck", decks.firstOrNull { it.id == deckID }?.name ?: "No decks", !running, {
                        decks.map { deck -> MenuEntry.Item(deck.name, checked = deck.id == deckID) { savedDeck = deck.id } }
                    }, "downloads.deck")
                }
                RowDivider()
                PickerRow("Image quality", quality.label, !running, {
                    ArtworkQuality.entries.map { value -> MenuEntry.Item(value.label, checked = value == quality) { qualityID = value.id } }
                }, "downloads.quality")
                if (downloadScope != "tokens") {
                    RowDivider()
                    IosToggle(includeTokens, { if (!running) includeTokens = it }, tint = BrandTheme.ember) {
                        Text("Include tokens", color = Color.White, style = sf(17f))
                    }
                }
            }
            FormSection("On this device") {
                if (loadingCatalogue && downloadScope == "catalogue") ProgressRow("Reading the installed catalogue…")
                if (catalogueError != null && downloadScope == "catalogue") Text(catalogueError, Modifier.padding(vertical = 10.dp), color = rgb(1.0, 0.27, 0.23), style = sf(17f))
                if (downloadScope != "tokens") {
                    Labeled("Cards", if (scanPending || (downloadScope == "catalogue" && loadingCatalogue)) "Checking needed"
                        else "${number.format(current!!.cards + current.faceStored)} / ${number.format(names.size + current.faceTotal)}",
                        Modifier.semantics { contentDescription = "downloads.cards" })
                    RowDivider()
                }
                Labeled("Tokens", if (scanPending || current?.tokensKnown != true && downloadsTokens) "Checking needed"
                    else "${number.format(current!!.tokenStored)} / ${number.format(current.tokenTotal)}")
                RowDivider()
                if (scanning) { ProgressRow("Checking local files…"); RowDivider() }
                Labeled("Stored", Formatter.formatFileSize(context, current?.bytes ?: Artwork.storedDownloadBytes(context)))
                RowDivider()
                Labeled("Missing artwork", missingSummary)
                RowDivider()
                Labeled("Estimated additional download", estimate)
                Text("Approximate at ${quality.label.lowercase()} quality. Actual download and device storage vary with image sizes, metadata and images already stored.",
                    Modifier.padding(bottom = 8.dp), color = secondary, style = sf(12f))
                RowDivider()
                FormButton("Check for missing artwork", !running && !scanning, "downloads.check") { runScan() }
            }
            FormSection("Download", "You can play or leave the app while images download. Wi-Fi is recommended.") {
                IosToggle(remoteArtwork, { enabled -> Artwork.setEnabled(context, enabled); if (!enabled && running) ArtworkDownloadService.pause(context) }, tint = BrandTheme.ember,
                    modifier = Modifier.semantics { contentDescription = "nativeArtwork.downloads" }) {
                    Text("Download card artwork", color = Color.White, style = sf(17f))
                }
                Text("Uses Scryfall. Online requests share your IP and card names, including your hand.", Modifier.padding(bottom = 8.dp), color = secondary, style = sf(12f))
                RowDivider()
                val progress = downloadState.progress
                if (running) {
                    val preparing = progress.status.startsWith("Preparing") || progress.status.startsWith("Checking")
                    if (preparing) ProgressRow("Preparing image list…")
                    else LinearProgressIndicator({ if (progress.total == 0) 0f else progress.completed.toFloat() / progress.total },
                        Modifier.fillMaxWidth().padding(vertical = 12.dp), color = BrandTheme.ember, trackColor = Color.White.copy(alpha = 0.15f))
                    Text(progress.status, Modifier.padding(bottom = 8.dp).semantics { contentDescription = "downloads.status" }, color = Color.White, style = sf(16f))
                    RowDivider()
                    FormButton("Cancel download", true, "downloads.cancel") { ArtworkDownloadService.pause(context) }
                } else {
                    if (progress.total > 0 && progress.status.isNotEmpty()) {
                        Text(progress.status, Modifier.padding(vertical = 10.dp).semantics { contentDescription = "downloads.status" }, color = Color.White, style = sf(16f))
                        RowDivider()
                    }
                    FormButton("Download missing artwork", remoteArtwork && (names.isNotEmpty() || downloadScope == "tokens") && !scanning &&
                        !(downloadScope == "catalogue" && loadingCatalogue), "downloads.start") { if (fullCatalogue) confirmFull = true else startDownload() }
                    if (ArtworkDownloadService.hasPending(context)) {
                        RowDivider()
                        FormButton("Resume previous download", remoteArtwork, "downloads.resume") { runCatching { ArtworkDownloadService.resume(context) } }
                    }
                }
            }
            FormSection {
                MoreInfo(engineReady, catalogueError == null)
            }
            val failures = downloadState.failures
            if (failures.isNotEmpty()) FormSection("Needs attention") {
                failures.take(20).forEach { Text(it, Modifier.padding(vertical = 6.dp), color = Color.White, style = sf(16f)) }
                if (failures.size > 20) Text("And ${failures.size - 20} more issues. Check missing artwork below.", color = secondary, style = sf(12f))
                Text("Use Download missing artwork to retry. Already stored artwork is preserved.", Modifier.padding(vertical = 6.dp), color = secondary, style = sf(12f))
            }
            if (current != null && current.missingTokens.isNotEmpty()) FormSection {
                Disclosure("Missing tokens · ${number.format(current.missingTokens.size)}") {
                    current.missingTokens.take(20).forEach { Text("Token $it", color = Color.White, style = sf(17f)) }
                }
            }
            if (current != null && current.missingCards.isNotEmpty() && downloadScope != "tokens") FormSection {
                Disclosure("Missing cards · ${number.format(current.missingCards.size)}") {
                    current.missingCards.take(20).forEach { Text(it, color = Color.White, style = sf(17f)) }
                    if (current.missingCards.size > 20) Text("And ${current.missingCards.size - 20} more", color = secondary, style = sf(17f))
                }
            }
        }
    }
    if (confirmFull) IosAlert("Download missing artwork?",
        "$missingSummary ${if (discoveryIncomplete) "The additional download estimate needs checking." else "Estimated additional download: $estimate."} This is approximate; actual download and device storage vary. Images continue downloading while you play or leave the app. Wi-Fi is recommended.",
        listOf("Download missing images · ${quality.label}" to { confirmFull = false; startDownload() }), cancel = { confirmFull = false })
}

@Composable
private fun ProgressRow(title: String) {
    Row(Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp).padding(vertical = 8.dp), verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp)) {
        CircularProgressIndicator(Modifier.size(18.dp), color = secondary, strokeWidth = 2.dp)
        Text(title, color = secondary, style = sf(17f))
    }
}

@Composable
private fun Disclosure(title: String, content: @Composable ColumnScope.() -> Unit) {
    var expanded by remember { mutableStateOf(false) }
    Column(Modifier.fillMaxWidth()) {
        Row(Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp).clickable { expanded = !expanded }, verticalAlignment = Alignment.CenterVertically) {
            Text(title, Modifier.weight(1f), color = Color.White, style = sf(16f))
            SfImage(if (expanded) "chevron.down" else "chevron.right", BrandTheme.ember, 14.dp)
        }
        if (expanded) Column(Modifier.padding(bottom = 10.dp), verticalArrangement = Arrangement.spacedBy(8.dp), content = content)
    }
}

@Composable
private fun MoreInfo(engineReady: Boolean, catalogueIncluded: Boolean) {
    Disclosure("More info") {
        @Composable fun label(text: String, ok: Boolean) = Row(horizontalArrangement = Arrangement.spacedBy(10.dp), verticalAlignment = Alignment.CenterVertically) {
            SfImage(if (ok) "checkmark.circle" else "exclamationmark.circle", BrandTheme.ember, 17.dp)
            Text(text, color = Color.White, style = sf(16f))
        }
        label(if (engineReady) "Local engine ready" else "Local engine not ready", engineReady)
        label(if (catalogueIncluded) "Card catalogue included" else "Card catalogue unavailable", catalogueIncluded)
        label("Mana symbols included", true)
        for (text in listOf(
            "These downloads supply artwork for decks and games. Rules and the supported card catalogue are already included; artwork is optional.",
            "Compact saves space. Standard balances clarity and size. High gives the sharpest inspection images. Higher-quality files already stored count toward lower-quality coverage.",
            "Full catalogue covers this build’s supported cards, not every printing. Alternate faces are checked during download. Estimates use currently discovered missing images; more faces or tokens may be found while preparing. Actual download and storage vary. Check for missing artwork after app updates.",
            "Full and token-only downloads use Scryfall’s bulk image index. Deck and on-demand requests share card names and your IP address. Stored artwork works offline.",
            "Compact is fastest. Downloads use several direct image transfers at once and remember completed files. Android shows a notification while it downloads; closing the app from recents may pause transfers until you reopen it. The initial image-list preparation may need the app open on a slow connection.",
            "Storage is capped at 20 GB, with 1 GB of free space reserved. Unavailable or ambiguous token art remains a labeled placeholder. Use Download missing artwork to retry interruptions.",
        )) Text(text, color = Color.White, style = sf(16f))
    }
}
